src/service/resources-impl.js

```diff
@@ -407,18 +410,54 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
+      return;
+    }
+    try {
+      this.isCurrentlyBuildingPendingResources_ = true;
+      this.buildReadyResourcesUnsafe_();
+    } finally {
+      this.isCurrentlyBuildingPendingResources_ = false;
+    }
+  }
+
+  /** @private */
+  buildReadyResourcesUnsafe_() {
     for (let i = 0; i < this.pendingBuildResources_.length; i++) {
       const resource = this.pendingBuildResources_[i];
-      if (this.documentReady_ || hasNextNodeInDocumentOrder(resource.element)) {
-        resource.build();
-        // Resource is built remove it from the pending list and step back
-        // one in the index to account for the removed item.
+      try {
```

Comment:
We shouldn't need another try catch (or another below it) here. Just splice before calling `#build`.